pkg/log/logshipper/logshipper.go

```diff
@@ -0,0 +1,127 @@
+package logshipper
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/sendbuffer"
+)
+
+// TODO: consolidate this var, but where?
+var observabilityIngestTokenKey = []byte("observability_ingest_auth_token")
+
+const (
+	truncatedFormatString = "%s[TRUNCATED]"
+)
+
+type LogShipper struct {
+	sender            *authedHttpSender
+	sendBuffer        *sendbuffer.SendBuffer
+	logger            log.Logger
+	knapsack          types.Knapsack
+	stopFunc          context.CancelFunc
+	isShippingEnabled bool
+}
+
+func New(k types.Knapsack) (*LogShipper, error) {
+	token, _ := k.TokenStore().Get(observabilityIngestTokenKey)
+
+	logEndpoint, err := logEndpoint(k)
+	if err != nil {
+		return nil, err
+	}
+
+	sender := newAuthHttpSender(logEndpoint, string(token))
+
+	sendInterval := time.Minute * 1
+	if k.Debug() {
+		sendInterval = time.Second * 1
+	}
+
+	sendBuffer := sendbuffer.New(sender, sendbuffer.WithSendInterval(sendInterval))
+	logger := log.NewJSONLogger(sendBuffer)
+
+	return &LogShipper{
+		sender:     sender,
+		sendBuffer: sendBuffer,
+		logger:     logger,
+		knapsack:   k,
+	}, nil
+}
+
+func (ls *LogShipper) Ping() {
+	token, _ := ls.knapsack.TokenStore().Get(observabilityIngestTokenKey)
+	ls.sender.authtoken = string(token)
+
+	endpoint, err := logEndpoint(ls.knapsack)
```

Comment:
Wouldn't you need to implement `FlagsChanged(flagKeys ...keys.FlagKey)` and call `knapsack.RegisterChangeObserver` to get agent flags (i.e. `LogIngestServerURL`) updates on time?

Reply:
I kinda took the lazy rout and just subscribed to all agent_flags updates, felt easier to just handle all the control server updates in Ping(), however if I'm willing to change it if you think it would be cleaner.

https://github.com/James-Pickett/launcher/blob/james/log-shipping/cmd/launcher/launcher.go#L324

Reply:
I think you're right that we should come up with some patterns for doing this consistently. But I don't know what it'll be yet.